packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", callback);
+
+      await subscription.on("error", error => {
+        errorFn();
+        subscription.removeAllListeners();
+
+        // FIXME: Assertion not working
         //expect(1).to.equal(2);
-      }
-    };
+        expect(error.message).to.match(/uncle/);
+      });
+
+      const snapshotId = await createSnapshot(simpleStorage.getProvider());
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
 
-    subscription.on("confirmation", onMessage);
+      expect(confirmationFn.called).to.be.true;
 
-    await mineBlocks(simpleStorage.getProvider(), 8);
+      await revertFromSnapshot(simpleStorage.getProvider(), snapshotId);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      // Broadcast same message again (simulate reorg)
+      subscription = simpleStorage.setValue("hello world");
+
+      expect(errorFn.called).to.be.true;
+    });
   });
 
-  it("should throw error when subscribing on invalid event", async () => {
-    const events = ["ValueChanged", "InvalidEvent"];
-    try {
+  describe("contract events subscriptions tests", async () => {
+    beforeEach("assign a wallet to the contract", () => {
+      expect(() => {
+        simpleStorage.setWallet(wallet);
+      }).to.not.throw();
+    });
+
+    it("should throw error when subscribing on invalid event", async () => {
+      const events = ["ValueChanged", "InvalidEvent"];
+
+      expect(() => {
+        subscription = simpleStorage.subscribe(events);
+      }).to.throw();
+    });
+
+    it("should throw error then listening on invalid event", async () => {
+      const events = ["ValueChanged"];
       const subscription = simpleStorage.subscribe(events);
-      assert(false, "subscription with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
 
-  it("should throw error then listening on invalid event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
+      expect(() => {
+        subscription.on("InvalidEvent", () => {});
+      }).to.throw();
+    });
 
-    try {
-      subscription.on("InvalidEvent", () => {});
-      assert(false, "listening with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
+    it("should listen to an event", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const fakeFn = sinon.fake();
 
-  it.skip("should listen to an event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
-    const handlerFunction = async message => {};
+      const handlerFunction = message => {
+        const { data } = message;
+        const { args } = data;
+        fakeFn();
+        subscription.removeListener("ValueChanged", handlerFunction);
+      };
+
+      subscription.on("ValueChanged", handlerFunction);
```

Comment:
Maybe it would be good to add a `subscription.on("error"` to this test too and make sure it isn't triggered. Alternatively, that could be a separate test.

Reply:
Resolved on PR https://github.com/tasitlabs/TasitSDK/pull/102

Reply:
Great! Resolving this thread then.